docs/proposals/diamond-design.md

```diff
@@ -0,0 +1,458 @@
+* This is useful information to trace back performance issues (e.g. one of the node took much more time than its siblings) or understand conflicts
+* I'd like to report about the status of the various splits and get some feed back about failure or completion
+* This is useful to make sure I can safely proceed with committing the work done, without necessarily requiring an exit status from the clients
+
+Design:
+* At commit time, we enrich the bundle metadata descriptor
+  * All started splits with their originating contributor, completion status and timings
+  * Bundle entries with their originating diamond and split IDs
+* This metadata is at first recorded in:
+  ```
+  /splits/{repo}/{diamond-id}/{split-id}/split.yaml
+  ```
+* At commit time, all split metadata are merged in some extra fields in the metadata of the bundle, accessible by using `datamon bundle get`.
+  Changes affect the content of:
+  ```
+  /bundles/{repo}/{bundle-id}/bundle.yaml
+  ```
+* This does not break existing bundles
+
+
+#### Lineage tracking
+
+Premises:
+* We assume that usually, all diamonds are initiated and run by the same contributor (e.g. some ARGO workflow with a runID)
+* However, in the case of several contributors, each contributing some splits to the diamond,
+  we should keep track of all of them in the list of contributors to that bundle
+
+Design:
+* Each split keeps track of its originating contributor
+* This metadata is at first recorded in:
+  ```
+  /splits/{repo}/{diamond-id}/{split-id}/split.yaml
+  ```
+* At commit time, we merge contributors into the resulting bundle:
+  ```
+  /bundles/{repo}/{bundle-id}/bundle.yaml
+  ```
+
+## Additional commands
+
+### List all initiated diamonds on a repo
+```
+datamon bundle diamond list --repo {repo}
+```
+
+Completed diamonds are available in this listing until they are eventually cleanup by the background cleaner.
+
+### List all ongoing splits for a diamond on a repo
+```
+datamon bundle diamond split list --repo {repo} --diamond {diamond-id}
+```
+
+Report about the ID and status and timings of every ongoing or failed split.
+
+### List files in one or all splits
+```
+datamon bundle diamond split files --repo {repo} --diamond {diamond-id} [--split-id {split-id}]
+```
+
+### Report about conflicting files on a bundle
+```
+datamon bundle get --repo {repo} --bundle {bundle-id} --with-conflicts|--with-checkpoints
+```
+This will report about conflicting paths, or checkpointed versions in the case of incremental uploads.
+
+### Report about diamond-originated information on a bundle
+```
+datamon bundle get --repo {repo} --bundle {bundle-id} --format `{{ range .Diamonds}}{{ printf "%v" .}}{{ end }}`
+```
+or
+```
+datamon bundle get --repo {repo} --bundle {bundle-id} --with-diamonds
+```
+
+### Inspect conflicts
+
+```
+datamon bundle list files --repo {repo} --bundle {bundle-id} --with-conflicts|--with-checkpoints
```

Comment:
Not in MVP... this might be useful or might be totally unused. For now, I still prefer the original proposal to put all the conflicting files in a special folder within the bundle.

Reply:
Handling conflicts is what we need to help developers avoid vs. spending efforts to make it easier to handle.
Also for the current use cases, we do not expect this. The conflict handling was added to at least detect it if it occurs, we do not need to solve this in an elegant manner.
We can write objects in the diamond object-store path for each file to be written and help clients get a clock on. file as an example of avoiding conflicts.. so in terms of putting in the time, I would rather avoid conflicts than making it easier to handle. Either way, we do not need to implement an extension to the bundle MD that is designed purely for conflict handling.

Reply:
Agree with the priority to release the full feature later. However we need to prepare this: if we lose the references to the hashes we won t be able to deliver it without yet another migration.
Storing all intermediary hashes (comflicts or checkpoints) will s what i am after

Reply:
Agree with the priority to release the full feature later. However we need to prepare this: if we lose the references to the hashes we won t be able to deliver it without yet another migration.
Storing all intermediary hashes (comflicts or checkpoints) will s what i am after

Reply:
Ok not to extend bundle metadata. Will provision another location on vmetadata for that

Reply:
Ok not to extend bundle metadata. Will provision another location on vmetadata for that

Reply:
Agreed with everything. Now conflicts are stored as special paths (note: filtered from uploads, like `.datamon` metadata files). Mentioned many things as "nice to have".
Kept however the distinction between conflict and checkpoints, which remains only matter of terminology.